src/nrniv/nrnapi.h

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include "hocdec.h"
+#include "section.h"
+
+// we define these here to allow the API to be independent of internal details
+#define STACK_IS_STR 1
+#define STACK_IS_VAR 2
+#define STACK_IS_NUM 3
+#define STACK_IS_OBJVAR 4
+#define STACK_IS_OBJTMP 5
+#define STACK_IS_USERINT 6
+#define STACK_IS_SYM 7
+#define STACK_IS_OBJUNREF 8
+#define STACK_UNKNOWN -1
+
+class SectionListIterator {
+public:
+  SectionListIterator(hoc_Item *);
+  Section *next(void);
+  int done(void);
+
+private:
+  hoc_Item *initial;
+  hoc_Item *current;
+};
+
+class SymbolTableIterator {
+public:
+  SymbolTableIterator(Symlist *);
+  char const *next(void);
+  int done(void);
+
+private:
+  Symbol *current;
+};
+
+extern "C" {
+/****************************************
+ * Initialization
+ ****************************************/
+int nrn_init(int argc, const char **argv);
+void nrn_redirect_stdout(int (*myprint)(int, char *));
+
+/****************************************
+ * Sections
+ ****************************************/
+Section *nrn_new_section(char const *const name);
+void nrn_connect_sections(Section *child_sec, double child_x,
+                          Section *parent_sec, double parent_x);
+void nrn_set_section_length(Section *sec, double length);
+double nrn_get_section_length(Section *sec);
```

Comment:
Agree, but see comment about `section_length` (which this function calls) not promising const `Section` (and indeed, it can update the datum fields associated with it).

Reply:
Added the const here, and `const_cast` it away.